### PR TITLE
Adding custom headers in batch operations

### PR DESCRIPTION
**Pull request by svetvasilev**
Introducing the capability of adding custom operation headers in batches. This seems to be rather widespread in SAP exposed OData endpoints where certain behaviour of the service is governed through such headers.
The solution is the introduction of a WithHeaders() method in the Fluent/IBoundClient interface. It takes an IDictionary of type string, string as parameter with the header:value kvps.
It then serializes the header kvps accordingly in the BatchWriter class within the corresponding batch operation. 
This comes very handy in situations where third party service has some customization OData through such headers.
In addition there was introduced a way to intercept the ODataMessage that represents the individual operation in a batch in order to allow for better testing. This way it is easier to verify that the headers have been properly added to the operation message in a non-intrusive manner.

### CODE REVIEW

File: src/Simple.OData.Client.Core/Fluent/CommandDetails.cs [hunk 1]
```diff
@@ -46,7 +47,7 @@ public CommandDetails(Session session, FluentCommand parent, ConcurrentDictionar
             this.SelectColumns = new List<string>();
             this.OrderbyColumns = new List<KeyValuePair<string, bool>>();
             this.MediaProperties = new List<string>();
-            this.BatchEntries = batchEntries;
+            this.BatchEntries = batchEntries;
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/CommandDetails.cs]:
```suggestion
            this.BatchEntries = batchEntries;
```

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs [hunk 2]
```diff
@@ -647,6 +656,10 @@ private string Format()
 
         internal string ActionName => _details.ActionName;
 
+        internal IDictionary<string, string> OperationHeaders => _details.OperationHeaders;
+
+        internal bool HasOperationHeaders => _details.OperationHeaders == null;
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/FluentCommand.cs]: It seems like this should be `_details.OperationHeaders != null`

File: src/Simple.OData.Client.Core/Fluent/IBoundClient.cs [hunk 3]
```diff
@@ -496,6 +496,12 @@ public interface IBoundClient<T> : IFluentClient<T, IBoundClient<T>>
         /// <param name="associationsToSetByValue">The list of associations to be passed by value for deep insert/update.</param>
         /// <returns>Self.</returns>
         IBoundClient<T> Set(T entry, Expression<Func<T, object>> associationsToSetByValue);
+        /// <summary>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/IBoundClient.cs]:
```suggestion

        /// <summary>
```

File: src/Simple.OData.Client.Core/Simple.OData.Client.Core.csproj [hunk 4]
```diff
@@ -5,6 +5,9 @@
     <RootNamespace>Simple.OData.Client</RootNamespace>
     <OutputPath>$(SolutionDir)/build/$(Configuration)/$(AssemblyName)/$(TargetFramework)/</OutputPath>
     <OutDir>$(OutputPath)</OutDir>
+    <SignAssembly>false</SignAssembly>
+    <DelaySign>false</DelaySign>
+    <AssemblyOriginatorKeyFile>C:\Projects\Simple.OData.Client\Key.snk</AssemblyOriginatorKeyFile>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Simple.OData.Client.Core.csproj]: Revert these changes.

File: src/Simple.OData.Client.Dynamic/Simple.OData.Client.Dynamic.csproj [hunk 5]
```diff
@@ -4,6 +4,9 @@
     <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
     <OutputPath>$(SolutionDir)/build/$(Configuration)/$(AssemblyName)/$(TargetFramework)/</OutputPath>
     <OutDir>$(OutputPath)</OutDir>
+    <SignAssembly>false</SignAssembly>
+    <AssemblyOriginatorKeyFile>C:\Projects\Simple.OData.Client\Key.snk</AssemblyOriginatorKeyFile>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Dynamic/Simple.OData.Client.Dynamic.csproj]: Revert these changes.

File: src/Simple.OData.Client.UnitTests/Simple.OData.Client.UnitTests.csproj [hunk 6]
```diff
@@ -7,6 +7,8 @@
     <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
     <OutputPath>$(SolutionDir)/build/$(Configuration)/$(AssemblyName)/$(TargetFramework)/</OutputPath>
     <OutDir>$(OutputPath)</OutDir>
+    <SignAssembly>false</SignAssembly>
+    <AssemblyOriginatorKeyFile>C:\Projects\Simple.OData.Client\Key.snk</AssemblyOriginatorKeyFile>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.UnitTests/Simple.OData.Client.UnitTests.csproj]: Revert these changes.

File: src/Simple.OData.Client.V3.Adapter/Simple.OData.Client.V3.Adapter.csproj [hunk 7]
```diff
@@ -4,7 +4,12 @@
     <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
     <OutputPath>$(SolutionDir)/build/$(Configuration)/$(AssemblyName)/$(TargetFramework)/</OutputPath>
     <OutDir>$(OutputPath)</OutDir>
+    <SignAssembly>false</SignAssembly>
+    <AssemblyOriginatorKeyFile>C:\Projects\Simple.OData.Client\Key.snk</AssemblyOriginatorKeyFile>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V3.Adapter/Simple.OData.Client.V3.Adapter.csproj]: Revert these changes.

File: src/Simple.OData.Client.V4.Adapter/Simple.OData.Client.V4.Adapter.csproj [hunk 8]
```diff
@@ -4,7 +4,12 @@
     <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
     <OutputPath>$(SolutionDir)/build/$(Configuration)/$(AssemblyName)/$(TargetFramework)/</OutputPath>
     <OutDir>$(OutputPath)</OutDir>
+    <SignAssembly>false</SignAssembly>
+    <AssemblyOriginatorKeyFile>C:\Projects\Simple.OData.Client\Key.snk</AssemblyOriginatorKeyFile>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V4.Adapter/Simple.OData.Client.V4.Adapter.csproj]: Revert these changes.

File: Simple.OData.Client.sln [hunk 9]
```diff
@@ -10,6 +10,7 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
 		build.ps1 = build.ps1
 		src\CommonAssemblyInfo.cs = src\CommonAssemblyInfo.cs
 		src\CommonAssemblyVersion.cs = src\CommonAssemblyVersion.cs
+		Key.snk = Key.snk
```
Comment (watfordgnf) [inline Simple.OData.Client.sln]: Revert this change.

File: Simple.OData.Client.sln [hunk 10]
```diff
@@ -26,25 +27,25 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Tests", "Tests", "{78D04AF9
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Simple.OData.NorthwindModel", "src\Simple.OData.NorthwindModel\Simple.OData.NorthwindModel.csproj", "{2A909E4D-6493-4D36-9FDE-786B89F9A387}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Simple.OData.Client.Core", "src\Simple.OData.Client.Core\Simple.OData.Client.Core.csproj", "{361868B8-A63E-4C74-AA87-B70E1AB22313}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Simple.OData.Client.Core", "src\Simple.OData.Client.Core\Simple.OData.Client.Core.csproj", "{361868B8-A63E-4C74-AA87-B70E1AB22313}"
```
Comment (watfordgnf) [inline Simple.OData.Client.sln]: Seems odd that these folder GUIDs changed. Any reason this change is necessary?

File: src/Simple.OData.Client.Core/Adapter/BatchWriterBase.cs [hunk 11]
```diff
@@ -86,7 +98,7 @@ public void MapContentId(IDictionary<string, object> entryData, string contentId
 
         public IDictionary<object, IDictionary<string, object>> BatchEntries { get; private set; }
 
-        public async Task<object> CreateOperationMessageAsync(Uri uri, string method, string collection, IDictionary<string, object> entryData, bool resultRequired)
+        public async Task<object> CreateOperationMessageAsync(Uri uri, string method, string collection, IDictionary<string, object> entryData, IDictionary<string, string> operationHeaders, bool resultRequired)
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Adapter/BatchWriterBase.cs]: This is a breaking change to a public class. A method should be left with the old signature that forwards to this new public signature (with the appropriate value for `operationHeaders`).

File: src/Simple.OData.Client.Core/Adapter/IBatchWriter.cs [hunk 12]
```diff
@@ -10,7 +10,7 @@ namespace Simple.OData.Client
     public interface IBatchWriter
     {
         Task<ODataRequest> CreateBatchRequestAsync(IODataClient client, IList<Func<IODataClient, Task>> actions, IList<int> responseIndexes);
-        Task<object> CreateOperationMessageAsync(Uri uri, string method, string collection, IDictionary<string, object> entryData, bool resultRequired);
+        Task<object> CreateOperationMessageAsync(Uri uri, string method, string collection, IDictionary<string, object> entryData, IDictionary<string, string> operationHeaders, bool resultRequired);
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Adapter/IBatchWriter.cs]: This is a breaking change to a public interface.

File: src/Simple.OData.Client.Core/Adapter/IRequestWriter.cs [hunk 13]
```diff
@@ -10,8 +10,8 @@ public interface IRequestWriter
     {
         Task<ODataRequest> CreateGetRequestAsync(string commandText, bool scalarResult);
         Task<ODataRequest> CreatePutRequestAsync(string commandText, Stream stream, string contentType, bool optimisticConcurrency);
-        Task<ODataRequest> CreateInsertRequestAsync(string collection, string commandText, IDictionary<string, object> entryData, bool resultRequired);
-        Task<ODataRequest> CreateUpdateRequestAsync(string collection, string entryIdent, IDictionary<string, object> entryKey, IDictionary<string, object> entryData, bool resultRequired);
+        Task<ODataRequest> CreateInsertRequestAsync(string collection, string commandText, IDictionary<string, object> entryData, IDictionary<string, string> operationHeaders, bool resultRequired);
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Adapter/IRequestWriter.cs]: This is a breaking change to a public interface.

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs [hunk 14]
```diff
@@ -188,6 +188,15 @@ public FluentCommand WithMedia(params ODataExpression[] properties)
             return WithMedia(properties.Select(x => x.Reference));
         }
 
+        public FluentCommand WithHeaders(IDictionary<string, string> operationHeaders)
+        {
+            if (IsBatchResponse)
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/FluentCommand.cs]: This should include a method level comment that headers are not applied per-request in a batch response.

File: src/Simple.OData.Client.IntegrationTests/Simple.OData.Client.IntegrationTests.csproj [hunk 15]
```diff
@@ -6,6 +6,8 @@
     <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
     <OutputPath>$(SolutionDir)/build/$(Configuration)/$(AssemblyName)/$(TargetFramework)/</OutputPath>
     <OutDir>$(OutputPath)</OutDir>
+    <SignAssembly>false</SignAssembly>
+    <AssemblyOriginatorKeyFile>C:\Projects\Simple.OData.Client\Key.snk</AssemblyOriginatorKeyFile>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.IntegrationTests/Simple.OData.Client.IntegrationTests.csproj]: Revert these changes.

File: src/Simple.OData.Client.V3.Adapter/RequestWriter.cs [hunk 16]
```diff
@@ -18,17 +18,44 @@ namespace Simple.OData.Client.V3.Adapter
     public class RequestWriter : RequestWriterBase
     {
         private readonly IEdmModel _model;
+        
+        private Func<IODataRequestMessage, ODataMessageWriterSettings, IEdmModel, ODataMessageWriter> _messageWriterInitializer;
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V3.Adapter/RequestWriter.cs]: This should be `readonly`.
```suggestion
        private readonly Func<IODataRequestMessage, ODataMessageWriterSettings, IEdmModel, ODataMessageWriter> _messageWriterInitializer;
```
This also looks like a good candidate for an interface, but that's up for debate.

File: src/Simple.OData.Client.V4.Adapter/RequestWriter.cs [hunk 17]
```diff
@@ -1,606 +1,640 @@
...
+
+namespace Simple.OData.Client.V4.Adapter
+{
+    public class RequestWriter : RequestWriterBase
+    {
+        private readonly IEdmModel _model;
+        private readonly Dictionary<ODataResource, ResourceProperties> _resourceEntryMap;
+        private readonly Dictionary<ODataResource, List<ODataResource>> _resourceEntries;
+
+        private Func<IODataRequestMessage, ODataMessageWriterSettings, IEdmModel, ODataMessageWriter> _messageWriterInitializer;
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V4.Adapter/RequestWriter.cs]:
```suggestion
        private readonly Func<IODataRequestMessage, ODataMessageWriterSettings, IEdmModel, ODataMessageWriter> _messageWriterInitializer;
```

File: src/Simple.OData.Client.Core/Fluent/CommandDetails.cs [hunk 18]
```diff
@@ -35,6 +35,7 @@ class CommandDetails
         public string MediaName { get; set; }
         public IEnumerable<string> MediaProperties { get; set; }
         public ConcurrentDictionary<object, IDictionary<string, object>> BatchEntries { get; set; }
+        public IDictionary<string,string> OperationHeaders { get; set; }
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/CommandDetails.cs]:
```suggestion
        public IDictionary<string, string> OperationHeaders { get; set; }
```

File: src/Simple.OData.Client.Core/Adapter/RequestWriterBase.cs [hunk 19]
```diff
@@ -85,11 +85,7 @@ public async Task<ODataRequest> CreateUpdateRequestAsync(string collection, stri
             updateMethod = _session.Settings.PreferredUpdateMethod == ODataUpdateMethod.Merge ? RestVerbs.Merge : updateMethod;
 
             var entryContent = await WriteEntryContentAsync(
-                updateMethod, collection, entryIdent, entryData, resultRequired).ConfigureAwait(false);
-
-
-
-
+                updateMethod, collection, entryIdent, entryData, operationHeaders ,resultRequired).ConfigureAwait(false);
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Adapter/RequestWriterBase.cs]:
```suggestion
                updateMethod, collection, entryIdent, entryData, operationHeaders, resultRequired).ConfigureAwait(false);
```

File: src/Simple.OData.Client.Core/Simple.OData.Client.Core.csproj (hunk 4 above)
Comment (svetvasilev) [reply src/Simple.OData.Client.Core/Simple.OData.Client.Core.csproj]: Will do. Was in discussion with Vagif about pros&cons with signing. In my particular scenario I had to sign the assemblies and obviously my cleanup was not that thurough...

File: src/Simple.OData.Client.Dynamic/Simple.OData.Client.Dynamic.csproj (hunk 5 above)
Comment (svetvasilev) [reply src/Simple.OData.Client.Dynamic/Simple.OData.Client.Dynamic.csproj]: Same as above :)

File: src/Simple.OData.Client.IntegrationTests/Simple.OData.Client.IntegrationTests.csproj (hunk 15 above)
Comment (svetvasilev) [reply src/Simple.OData.Client.IntegrationTests/Simple.OData.Client.IntegrationTests.csproj]: Here as well :)

File: Simple.OData.Client.sln (hunk 9 above)
Comment (svetvasilev) [reply Simple.OData.Client.sln]: Hmmm..even this one sneaked..will do :)

File: Simple.OData.Client.sln (hunk 10 above)
Comment (svetvasilev) [reply Simple.OData.Client.sln]: Seems odd really...my VS went berserk on the project refs...

File: src/Simple.OData.Client.Core/Adapter/BatchWriterBase.cs (hunk 11 above)
Comment (svetvasilev) [reply src/Simple.OData.Client.Core/Adapter/BatchWriterBase.cs]: I spent actually quite some time with the suggested approach but didnt really find a good solution to this. The primary reason being that this method is invoked from other methods up in the call chain. And since the operationHeaders is sent all the way from the command that seemed to be the most viable options. 
If you take a closer look i have therefore updated all the references of this method with a null value in the places where it is being called. Unit tests updated accordingly.
This of course given the fact that i was just introduced to this lib. Hence I rely on some more input from someone who knows the lib inside out :)

File: src/Simple.OData.Client.Core/Adapter/IBatchWriter.cs (hunk 12 above)
Comment (svetvasilev) [reply src/Simple.OData.Client.Core/Adapter/IBatchWriter.cs]: Same as above

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs (hunk 14 above)
Comment (svetvasilev) [reply src/Simple.OData.Client.Core/Fluent/FluentCommand.cs]: Will do

**Comment by svetvasilev** [comment]
I was finally able to push all the request changes. Please review then accordingly and come back to me in case i should look at anything further.

**Comment by watfordgnf** [comment]
@svetvasilev there are some conflicts in `RequestWriter.cs` that need addressing.

**Comment by svetvasilev** [comment]
@watfordgnf - i just pushed the required changes, accompanied with fixed tests, to the branch. Could you have a second look at our convenience?
Thanks, Svet

**Comment by object** [comment]
The project source code is under significant changes (see discussion here https://github.com/simple-odata-client/Simple.OData.Client/discussions/840), so if the PR is still valid, please resumbit it after the modernization activity is complete.

**Comment by bcasari-orbit** [comment]
@object @watfordgnf When is this commit going to be merged and released? It contains some fundamental changes for using the Batch functionality. At the moment it is impossible to send a header for each individual operation in a Batch. This is important for example when making conditional operations in a Batch with the MS Dataverse API.

**Comment by swidz** [comment]
Same question as @bcasari-orbit .
In my case I need to set individual headers for internal requests inside batch to control custom plugin execution. The batch header cannot control this. Same, MS Dataverse API requirement.
